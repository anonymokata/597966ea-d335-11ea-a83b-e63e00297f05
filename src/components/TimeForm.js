import React from 'react';
import moment from 'moment';

import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';

const useStyles = makeStyles((theme) => ({
  container: {
    display: 'flex',
    flexWrap: 'wrap',
  },
  textField: {
    marginLeft: theme.spacing(1),
    marginRight: theme.spacing(1),
    width: 200,
  },
}));

const startOfDay = moment({
  'hour': 17
})

const endOfDay = moment(startOfDay).add(11, 'hours');

export default class TimeForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      startTime: startOfDay.format('HH:mm'), // default to minimum babysit time
      endTime: endOfDay.format('HH:mm'), // default to 5 hours later
      duration: 5
    }
  }
  handleStartChange = (e) => {
    const startTimeValue = e.target.value;
    const startTimeMoment = moment(startTimeValue, 'HH:mm');
    const endTimeMoment = moment(this.state.endTime, 'HH:mm');
    if(endTimeMoment.hour() < 12) {
      endTimeMoment.add(24, 'hours');
    }
    if(startTimeMoment.isBefore(startOfDay) ||
      startTimeMoment.isSameOrAfter(endTimeMoment) ||
      startTimeMoment.isAfter(endOfDay)
    )
    {
      // do nothing
    } else {
      this.setState({
        startTime: startTimeMoment.format('HH:mm')
      });
    }
  }
  handleEndChange = (e) => {
    const endTimeValue = e.target.value;
    let endTimeMoment = moment(endTimeValue, 'HH:mm');
    endTimeMoment = this.checkForNextDay(endTimeMoment);
    const startTimeMoment = moment(this.state.startTime, 'HH:mm');

    if (endTimeMoment.isBefore(startOfDay) ||
      endTimeMoment.isSameOrBefore(startTimeMoment) ||
      endTimeMoment.isAfter(endOfDay)
    ) {
      // do nothing
    } else {
      this.setState({
        endTime: endTimeMoment.format('HH:mm')
      });
    }
  }

  calculateDuration = () => {
    const startTimeMoment = moment(this.state.startTime, 'HH:mm');
    let endTimeMoment = moment(this.state.endTime, 'HH:mm');
    endTimeMoment = this.checkForNextDay(endTimeMoment);
    const duration = moment.duration(endTimeMoment.diff(startTimeMoment)).asMinutes();
    this.setState({
      duration
    })
  }

  checkForNextDay = (argMoment) => {
    if (argMoment.hour() < 12) {
      return argMoment.day(startOfDay.day()).add(1, 'day');
    } else {
      return argMoment.day(startOfDay.day());
    }
  }

  render() {
    return(
      <div>
        <form noValidate>
          <TextField
            id="startTime"
            label="Start Time"
            type="time"
            InputLabelProps={{
              shrink: true,
            }}
            inputProps={{
              step: 300, // 5 min
            }}
            onChange={this.handleStartChange}
            value={this.state.startTime}
          />
          <br />
          <TextField
            id="endTime"
            label="End Time"
            type="time"
            InputLabelProps={{
              shrink: true,
            }}
            inputProps={{
              step: 300, // 5 min
            }}
            onChange={this.handleEndChange}
            value={this.state.endTime}
          />
        </form>
        <button
          onClick={this.calculateDuration}
        >Update time:</button><p>{this.state.duration}</p>
      </div>
    )
  }
};