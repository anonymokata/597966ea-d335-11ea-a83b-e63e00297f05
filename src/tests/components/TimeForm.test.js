import React from 'react';
import { shallow } from 'enzyme';

import TimeForm from '../../components/TimeForm';
import { duration } from 'moment';

test('Should correctly render TimeForm', () => {
  const wrapper = shallow(<TimeForm />);
  expect(wrapper).toMatchSnapshot();
});

test('Should initialize form with start time 5:00pm and end time 04:00am', () => {
  const wrapper = shallow(<TimeForm />);
  expect(wrapper.state('startTime')).toBe('17:00');
  expect(wrapper.state('endTime')).toBe('04:00');
});