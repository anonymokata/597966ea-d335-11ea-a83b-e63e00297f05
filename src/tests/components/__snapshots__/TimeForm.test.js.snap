// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render TimeForm 1`] = `
<div>
  <form
    noValidate={true}
  >
    <[object Object]
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      id="startTime"
      inputProps={
        Object {
          "step": 300,
        }
      }
      label="Start Time"
      onChange={[Function]}
      type="time"
      value="17:00"
    />
    <br />
    <[object Object]
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      id="endTime"
      inputProps={
        Object {
          "step": 300,
        }
      }
      label="End Time"
      onChange={[Function]}
      type="time"
      value="04:00"
    />
  </form>
  <button
    onClick={[Function]}
  >
    Update time:
  </button>
  <p>
    660
  </p>
</div>
`;
